### Try Model: Tensor Flow Neural Network

In [1]:
from Starts.startmodskl import *
from Starts.startmodtf import *

local_kwargs {'drop_obj_col': False, 'data_path_2': '2_Titanic/data/titanic_no_nan.csv', 'data_path_3': '2_Titanic/data/test.csv', 'nan_mean_neighbors': False, 'nan_zero': False, 'nan_mean': True, 'nan_drop_row': False, 'data_path_1': '2_Titanic/data/train.csv', 'nan_drop_col': False}


Using TensorFlow backend.


In [2]:
nonan_data.head()

Survived  Pclass  Sex   Age  Ticket     Fare  Title  FamilySize
0         0       3    1  49.0     574   0.0000      5           0
1         1       2    0   7.0     571  26.2500      4           2
2         0       3    1  33.0     378   8.6625      5           0
3         0       3    1  24.0     519  24.1500      5           2
4         1       1    0  30.0       0  86.5000      4           0

In [3]:
stf = StartModTF(nonan_data)
# Try method Keras Tensorflow to predict the Survivors (column 'Survived')
model, y_true, y_pred = stf.keras_sequential(nonan_data, 'Survived', hidden_layers=2)

Epoch 1/1
712/712 [==============================] - 0s 623us/step - loss: 0.6720 - acc: 0.6096


In [4]:
# Convert predicted value by rounding it (>0.5 will be considered as 1 else 0)
y_pred = np.array([1 if item>0.5 else 0 for item in y_pred])

# Compare the predicted value with true_value
len(y_true), len(y_pred)

(179, 179)

In [5]:
confusion_matrix(y_true, y_pred)

array([[115,   0],
       [ 64,   0]])

### Result: 
#### the correct predicted result 115
#### the wrong predicted result 64

In [6]:
# update parameters
new_param={'hidden_units':[10,10,10], 'optimizer':'Adam', 'activation_fn':'sigmoid', 'learning_rate': 0.0001,
           'steps':2000, 'batch_size':10, 'num_epochs':10, 'feature_scl':True}
stf.update_parameters=new_param
model, y_true, y_pred = stf.keras_sequential(nonan_data, 'Survived', hidden_layers=2)

Epoch 1/10
712/712 [==============================] - 0s 456us/step - loss: 0.6810 - acc: 0.6081
Epoch 2/10
712/712 [==============================] - 0s 317us/step - loss: 0.6432 - acc: 0.6180
Epoch 3/10
712/712 [==============================] - 0s 320us/step - loss: 0.6239 - acc: 0.6896
Epoch 4/10
712/712 [==============================] - 0s 383us/step - loss: 0.6096 - acc: 0.6966
Epoch 5/10
712/712 [==============================] - 0s 390us/step - loss: 0.5938 - acc: 0.7135
Epoch 6/10
712/712 [==============================] - 0s 402us/step - loss: 0.5919 - acc: 0.7135
Epoch 7/10
712/712 [==============================] - 0s 206us/step - loss: 0.5720 - acc: 0.7331
Epoch 8/10
712/712 [==============================] - 0s 262us/step - loss: 0.5892 - acc: 0.6994
Epoch 9/10
712/712 [==============================] - 0s 278us/step - loss: 0.5631 - acc: 0.7233
Epoch 10/10
712/712 [==============================] - 0s 266us/step - loss: 0.5517 - acc: 0.7191


In [7]:
# Convert predicted value by rounding it (>0.5 will be considered as 1 else 0)
y_pred = np.array([1 if item>0.5 else 0 for item in y_pred])
confusion_matrix(y_true, y_pred)

array([[105,  10],
       [ 32,  32]])

### Result: 
#### the correct predicted result 177 -> 137
#### the wrong predicted result 2 -> 42 

### Result has been changed (getting worse) without any impact from code, so data preprocessing has major impact on final result. In our case, the feature 'Ticket' was not removed during preprocess phase (NaN processing) and included into data. 

### Conclusion: unnecessary feature might be removed before building model.